### PR TITLE
refactor: SBOM tool generation refactor

### PR DESCRIPTION
**Pull request by amanda-tarafa**
@jskeet as always one commit at a time. I'll all refactoring in a single commit before merging.

### CODE REVIEW

File: tools/Google.Cloud.Tools.SbomGenerator.Tests/Google.Cloud.Tools.SbomGenerator.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      <PackageReference Include="xunit" Version="2.5.3" />
+      <PackageReference Include="xunit.runner.visualstudio" Version="2.5.3" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Google.Cloud.Tools.SbomGenerator\Google.Cloud.Tools.SbomGenerator.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="GoldenTestData\Expected-Google.Cloud.OsLogin.V1Beta.3.0.0-beta05.nupkg-sbom.spdx.json">
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.SbomGenerator.Tests/Google.Cloud.Tools.SbomGenerator.Tests.csproj:22]: Possibly just use `Update="GoldenTestData\*"` for simplicity?

File: tools/Google.Cloud.Tools.SbomGenerator/Program.cs [hunk 2]
```diff
@@ -132,7 +132,7 @@ private static Spdx SpdxFrom(string nupkgPath)
     {
         // Compute the SHA-256 hash of the .nupkg file.
         using var sha256 = SHA256.Create();
-        byte[]? hash = null;
+        byte[] hash = null;
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.SbomGenerator/Program.cs]: We don't need to initialize this at all... we don't read from it until after the `using` statement which it's assigned in.

File: tools/Google.Cloud.Tools.SbomGenerator/Program.cs [hunk 3]
```diff
@@ -164,23 +164,11 @@ private static Spdx SpdxFrom(string nupkgPath)
...
-        return new Nuspec
-        {
-            Id = metadata?.SelectSingleNode("id")?.InnerText ?? "",
-            Version = metadata?.SelectSingleNode("version")?.InnerText ?? "",
-            Description = metadata?.SelectSingleNode("description")?.InnerText
-                ?? "",
-        };
+        XmlSerializer xmlSerializer = new XmlSerializer(typeof(NuspecPackage));
+
+        return xmlSerializer.Deserialize(nuspecXmlStream) as NuspecPackage;
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.SbomGenerator/Program.cs]: I generally prefer casting over `as`, as it has a better failure model. Probably not worth messing with the commit history though...

File: tools/Google.Cloud.Tools.SbomGenerator/Program.cs [hunk 4]
```diff
@@ -37,25 +37,23 @@ public static void Main(string[] args)
...
+        await WriteSpdxAsync(nupkgPath);
     }
 
     /// <summary>
     /// Examines a .nupkg file and writes an SPDX SBOM json file next to it.
     /// </summary>
     /// <param name="nupkgPath">Path to .nupkg file.</param>
-    /// <returns>Path to generated SPDX SBOM json file.</returns>
-    private static void WriteSpdx(string nupkgPath)
+    private static async Task WriteSpdxAsync(string nupkgPath)
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.SbomGenerator/Program.cs:50]: Any particular reason for going to an async model? It feels like it's making things a little more complex without a benefit - but I'm probably missing something.

File: tools/Google.Cloud.Tools.SbomGenerator/Program.cs [hunk 5]
```diff
@@ -130,11 +115,26 @@ private static Spdx SpdxFrom(string nupkgPath)
...
+        using var sha256 = SHA256.Create();
+        using var file = File.OpenRead(nupkgPath);
+
+        var hashBytes = await sha256.ComputeHashAsync(file);
+
+        // Use the SHA-256 hash to create a deterministic uuid.
+        var uuid = Uuid5FromHash(hashBytes);
 
-        return xmlSerializer.Deserialize(nuspecXmlStream) as NuspecPackage;
+        return (BitConverter.ToString(hashBytes).Replace("-", ""), uuid.ToString());
```
Comment (jskeet) [inline tools/Google.Cloud.Tools.SbomGenerator/Program.cs:142]: Remove `()` surrounding the expression? (Can do that later of course.)

**Review by jskeet — APPROVED** [review]
A few very minor comments, but nothing blocking.

File: tools/Google.Cloud.Tools.SbomGenerator/Program.cs (hunk 5 above)
Comment (amanda-tarafa) [reply tools/Google.Cloud.Tools.SbomGenerator/Program.cs:142]: It's a tuple that I'm returnig here, the hash text, and the uuid built from the hash bytes.

**Review by amanda-tarafa — COMMENTED** [review]
Review comments addressed in the last two commits.

If all looks good, I'll squash in two commits and merge.

**Pull request merged**